### ISSUE TITLE
Segmentation edges are not smooth

### ISSUE THREAD

**Issue by PanJinquan** [ultralytics/ultralytics#24272]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

I have found that the YOLO instance segmentation model (e.g., yolov8n-seg.pt and other instance segmentation models) may produce segmentation contours that protrude beyond the bounding box. I have localized the issue to the process_mask post-processing function.

<img width="571" height="759" alt="Image" src="https://github.com/user-attachments/assets/3a96040c-b9ec-4022-8f65-a7e87fa32056" />

Here is the original `ultralytics.utils.ops.process_mask` post-processing function:
```
 def process_mask(protos, masks_in, bboxes, shape, upsample: bool = False):
    c, mh, mw = protos.shape  # CHW
    masks = (masks_in @ protos.float().view(c, -1)).view(-1, mh, mw)  # NHW

    width_ratio = mw / shape[1]
    height_ratio = mh / shape[0]
    ratios = torch.tensor([[width_ratio, height_ratio, width_ratio, height_ratio]], device=bboxes.device)

    masks = crop_mask(masks, boxes=bboxes * ratios)  # NHW
    if upsample:
        masks = F.interpolate(masks[None], shape, mode="bilinear")[0]  # NHW
    return masks.gt_(0.0).byte()
```
The issue arises because the original algorithm first applies crop_mask and then upsamples (which may interpolate other values), leading to binarization and causing segmentation contours to possibly extend beyond the bounding box.

The improved method is to upsample first and then apply crop_mask:
```
def process_mask(protos, masks_in, bboxes, shape, upsample: bool = False):
    c, mh, mw = protos.shape  # CHW
    masks = (masks_in @ protos.float().view(c, -1)).view(-1, mh, mw)  # NHW

    width_ratio = mw / shape[1]
    height_ratio = mh / shape[0]
    ratios = torch.tensor([[width_ratio, height_ratio, width_ratio, height_ratio]], device=bboxes.device)

    # masks = crop_mask(masks, boxes=bboxes * ratios)  # NHW
    if upsample:
        masks = F.interpolate(masks[None], shape, mode="bilinear")[0]  # NHW
    masks = crop_mask(masks, boxes=bboxes)  # 
    return masks.gt_(0.0).byte()
```
<img width="569" height="759" alt="Image" src="https://github.com/user-attachments/assets/579b420b-c239-480c-8b35-89fb314c6010" />

### Environment

Ultralytics 8.4.23 🚀 Python-3.10.0 torch-2.10.0+cu128 CUDA:0 (NVIDIA GeForce RTX 3070, 7967MiB)
Setup complete ✅ (16 CPUs, 31.2 GB RAM, 96.7/373.8 GB disk)

OS                     Linux-6.12.63-amd64-desktop-rolling-x86_64-with-glibc2.38
Environment            Linux
Python                 3.10.0
Install                pip
Path                   /home/PKing/miniconda3/envs/py310/lib/python3.10/site-packages/ultralytics
RAM                    31.16 GB
Disk                   96.7/373.8 GB
CPU                    Intel Core(TM) i7-10700K 3.80GHz
CPU count              16
GPU                    NVIDIA GeForce RTX 3070, 7967MiB
GPU count              1
CUDA                   12.8

numpy                  ✅ 1.24.3>=1.23.0
matplotlib             ✅ 3.10.7>=3.3.0
opencv-python          ✅ 4.13.0.92>=4.6.0
pillow                 ✅ 12.1.1>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.15.3>=1.4.1
torch                  ✅ 2.10.0>=1.8.0
torch                  ✅ 2.10.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.25.0>=0.9.0
psutil                 ✅ 7.1.1>=5.8.0
polars                 ✅ 1.39.2>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18

### Minimal Reproducible Example

```
    model_file = "yolov8n-seg.pt"
    model = YOLO(model_file)
    results = model.predict('bus.jpg', imgsz=640, save=True)  
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @PanJinquan, thank you for the detailed bug report and for sharing your investigation into the segmentation mask post-processing behavior 🚀. This is an automated response to help move things forward quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this is a 🐛 Bug Report, thanks for already including a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) and environment details — that is very helpful 🙌

If possible, please also verify the issue persists after upgrading to the latest package version and retesting in a clean supported environment.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
You can achieve this by using `retina_masks=True` during prediction. The reason it's not default is because it's slower than than `process_mask`.

**Comment by wjddnwp29** [comment]
Hi @PanJinquan — since this had been quiet for a few weeks I went ahead and put together a fix (PR #24463) that reorders `process_mask` to upsample first and crop at full resolution, matching the behavior already used in `process_mask_native`. The `upsample=False` path is unchanged for backward compatibility, and I've added a regression test in `tests/test_python.py` that catches the leakage. Happy to step aside if you'd rather take this — please let me know!

**Comment by onuralpszr** [comment]
@wjddnwp29 @plashchynski did you guys tested with retina as well or still same ?

**Comment by wjddnwp29** [comment]
@onuralpszr 
Yes, tested both paths on bus.jpg with imgsz=640, yolov8n-seg.pt on this PR branch (commit 09bf030).

  At identical resolution (model input 640×640). Feeding the same protos / mask coefficients / bboxes through (a) the pre-PR
  process_mask, (b) the patched process_mask, and (c) process_mask_native (the function retina_masks=True calls):

  - mean mask-IoU(a vs c) = 0.9437
  - mean mask-IoU(b vs c) = 1.0000, and torch.equal(b, c) is True — bit-identical for every detection.
  - pixels outside predicted bbox: a=1135, b=0, c=0.
  - per-detection IoU(a vs c) ranges 0.86–0.99; tightest masks (the small/edge people in bus.jpg) drift the most under the old code,
  which matches @PanJinquan's screenshot.

  End-to-end (predict(retina_masks=False) vs predict(retina_masks=True) on this branch).

  - fast path: (6, 640, 480) masks, retina: (6, 1080, 810).
  - per-detection IoU: [0.978, 0.981, 0.983, 0.992, 0.980, 0.925], mean 0.973.
  - bbox-leak px: fast=7, retina=0.

  The remaining ~3% gap and 7 leak pixels are not from the crop order — they are the unavoidable cost of bilinear-upsampling at model
  resolution vs. retina's full-image resolution. Crop-order-wise the two paths are now equivalent.

  The retina path itself is untouched by this PR (process_mask_native was not modified, the upsample=False branch of process_mask was
  kept for backward compatibility), so existing retina users are unaffected.

  Happy to push the two verification scripts (verify_24272.py for the same-resolution head-to-head, verify_24272_e2e.py for the
  predict()-level comparison) to the PR if useful.
